include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+
+    for (unsigned int cell = 0; cell < cell_data.cells.size(); ++cell)
+      {
+        solution_values.resize(fe_coarse->n_dofs_per_cell());
+
+        // gather and resolve constraints
+        internal::VectorReader<Number, VectorizedArrayType> reader;
+        constraint_info.read_write_operation(
+          reader,
+          src,
+          reinterpret_cast<VectorizedArrayType *>(solution_values.data()),
+          cell,
+          1,
+          solution_values.size(),
+          true);
+
+        // evaluate and scatter
+        evaluator.reinit(cell);
+
+        evaluator.evaluate(solution_values, dealii::EvaluationFlags::values);
+
+        for (const auto q : evaluator.quadrature_point_indices())
+          values[q + cell_data.reference_point_ptrs[cell]] =
+            evaluator.get_value(q);
+      }
+  };
+
+  rpe.template evaluate_and_process<Number>(evaluation_point_results,
+                                            buffer,
+                                            evaluation_function);
+
+  // Weight operator in case some points are owned by multiple cells.
+  if (rpe.is_map_unique() == false)
+    {
+      const auto evaluation_point_results_temp = evaluation_point_results;
+      evaluation_point_results.assign(rpe.get_point_ptrs().size() - 1, 0);
+
+      const auto &ptr = rpe.get_point_ptrs();
+
+      for (unsigned int i = 0; i < ptr.size() - 1; ++i)
+        {
+          const auto n_entries = ptr[i + 1] - ptr[i];
+          if (n_entries == 0)
+            continue;
+
+          Number result = 0.0;
+
+          for (unsigned int j = 0; j < n_entries; ++j)
+            result += evaluation_point_results_temp[ptr[i] + j];
+          result /= n_entries;
+
+          evaluation_point_results[i] = result;
+        }
+    }
+
+  for (unsigned int j = 0; j < evaluation_point_results.size(); ++j)
+    dst.local_element(this->level_dof_indices_fine[j]) +=
+      evaluation_point_results[j];
+}
+
+
+
+template <int dim, typename Number>
+void
+MGTwoLevelTransferNonNested<dim, LinearAlgebra::distributed::Vector<Number>>::
+  restrict_and_add_internal(
+    LinearAlgebra::distributed::Vector<Number> &      dst,
+    const LinearAlgebra::distributed::Vector<Number> &src) const
+{
+  std::vector<Number> evaluation_point_results;
+  std::vector<Number> buffer;
+
+  evaluation_point_results.resize(rpe.get_point_ptrs().size() - 1);
+
+  for (unsigned int j = 0; j < evaluation_point_results.size(); ++j)
+    evaluation_point_results[j] =
+      src.local_element(this->level_dof_indices_fine[j]);
+
+  // Weight operator in case some points are owned by multiple cells.
+  if (rpe.is_map_unique() == false)
+    {
+      const auto &ptr = rpe.get_point_ptrs();
+
+      for (unsigned int i = 0; i < ptr.size() - 1; ++i)
+        {
+          const auto n_entries = ptr[i + 1] - ptr[i];
+          if (n_entries == 0)
+            continue;
+
+          evaluation_point_results[i] /= n_entries;
```

Comment:
Is this really the transpose of the function above? This would mean that the `process_and_evaluate` function works asymmetrically, because in case `n_entries` is larger than 1 somewhere, the lengths of `evaluation_points_results` here is different from the `evaluation_points_results` used in `prolongate_and_add`. Note that the index used here is `i`, whereas it is `j` (aka `ptr[i] + ...`) above. I am confused.

Reply:
Let me answer this one. Indeed the implementation in `RPE` not completely symmetric, which is related to the application cases for which each function was written.

The function `evaluate_and_process()` collect for each point the contributions from all cells.

However, `process_and_evaluate()` sends the data at the point to the relevant cells. The same data is sent to all cells. The copying is done here:

https://github.com/dealii/dealii/blob/09cb1410a77d398fd711cbcf0f9cb07e960c323b/include/deal.II/base/mpi_remote_point_evaluation.h#L467-L475

`evaluation_point_results[i] /= n_entries;` followed by the copying is as far as I see the transpose of averaging, which is done during prolongation.